README.txt

---------------------------------------------------------------------------

Release:      pykaraoke v0.7.5
Date:         26/08/2011
Author:       Kelvin Lawson <kelvinl@users.sourceforge.net>
License:      LGPL
Website:      http://www.kibosh.org/pykaraoke/
Contributors: John Schneiderman <JohnMS@member.fsf.org>
              William Ferrell <willfe@gmail.com>
              David Rose <pykar@ddrose.com>

---------------------------------------------------------------------------

INTRODUCTION

PyKaraoke is a karaoke player for Linux, FreeBSD, Windows and GP2X.

The following song formats are supported:
 * CDG (MP3+G, OGG+G)
 * MIDI/KAR
 * MPEG/AVI/Other video formats

MPEG2 files can usually be played directly within the PyKaraoke
framework.  Other file formats like AVI or QuickTime, or files which
use more exotic codecs like XviD, can also be handled by PyKaraoke,
but it will require an external program, such as mplayer or Windows
Media Player to show them.  This is a little less tightly integrated,
but it does work, and allows you to pick all your karaoke songs from
one database.

No song files are provided - this package provides you with the player
needed to play your own karaoke song files.

---------------------------------------------------------------------------

INSTALLATION (LINUX, SOURCE INSTALLS)

PyKaraoke requires the following libraries to be installed:

 * Python (www.python.org)
 * Pygame (www.pygame.org)
 * WxPython (www.wxpython.org)
 * SDL source distribution (www.libsdl.org)

Optional libraries are as follows:

 * Mutagen (http://code.google.com/p/mutagen/)
 * PyMedia - Only required for converting CDG/KAR to MPG

PyKaraoke now offers two builds: an ultra-portable Python version, and
a highly optimised C version. The default install is to use the faster
C version which requires the SDL source distribution for compilation
purposes. All supported platforms should support compilation of the C
version, but the portable Python-only version will continue to be
supported. Python-only installs do not require the SDL source
distribution, and instead require the NumPy Python module
(numpy.scipy.org).

If these libraries are not already installed on your system, you can
download them from the websites listed.

Linux users may find these packages are available directly from their
distro's package manager. For example Debian users can install all
prerequisites using:
	# apt-get install python-dev python-pygame libwxgtk-python 
                  libsdl-dev python-mutagen

With the prerequisites installed, unzip the release and run the following
as root:

	# python setup.py install

This installs the executables into /usr/bin, and you can then run
PyKaraoke from anywhere using "pykaraoke".

Alternatively you can run PyKaraoke without installing by simply
unzipping and running "python pykaraoke.py" from the unzip location.
Beware that the uninstalled version method requires the NumPy (or the
older Numeric) library, and runs the more portable but slower
Python-only version of the CDG player.

---------------------------------------------------------------------------

INSTALLATION (OS X)

PyKaraoke is known to work well on Macintosh OS X.  It has been tested
on OS X versions 10.3, 10.4, and 10.5.  On this platform, you must
install from source, similar to the Linux platforms, as described
above, since we do not presently provide a binary OS X installation.

Unfortunately, there are several supporting packages you must install
first in order to run PyKaraoke on OS X.

The biggest troublemaker in the mix is Pygame.  Pygame is available
via fink, MacPorts, and via precompiled versions.  If you are already
familiar with using fink or MacPorts, you may find one of these the
easiest way to install it.  However, we will leave this as your own
exercise; we describe below the process of installing it via the
precompiled version.

Unfortunately, installing the precompiled version of Pygame also
requires installing the latest version of Python from python.org,
instead of relying on the version of Python that comes pre-installed
with OS X.  This means you must visit:

http://www.pygame.org/download.shtml

  Follow the link for "Pygame OSX PackageManager Site".  Verify the
  version of Python that the binaries here are compiled for.  At the
  time of this writing, this was python version 2.5.

http://www.python.org/download/

  Follow the link for "Python 2.5.2 for Macintosh OS X", or whatever
  the version requested by pygame.org.  Note that the minor version
  (the final .2 in this example) doesn't matter; just get whichever
  minor version is offered that matches the major version required by
  pygame.org.

http://www.pygame.org/download.shtml

  Once you have installed the proper version of Python, go back to
  this site and install both pygame and pyobjc.

http://www.wxpython.org/

  Even though wxPython was already installed for your system default
  version of Python, you must now install a new version of wxPython to
  go with the custom version of Python you just installed.  Scroll
  down to "OS X Binaries" and install the latest version of wxPython
  runtime under the version of Python that you installed (e.g. under
  the Python 2.5 column).  We recommend picking the osx-unicode
  version, though this doesn't really matter much.

That will be all you need to install in order to play .kar files.  In
order to play .cdg files as well, you will also need to install the
NumPy package:

http://www.scipy.org/Download


Now you have everything installed that you need in order to play .kar
and .cdg files.  In order to launch PyKaraoke, open a terminal window
and type the command:

/Library/Frameworks/Python.framework/Versions/Current/bin/pythonw pykaraoke.py

You can use Platypus ( http://www.sveinbjorn.org/platypus ) to hide
this ugly command inside a convenient application icon, so you don't
have to type it every time.  For now, just type the command exactly as
shown above.  This accesses the version of Python that you have just
installed, as opposed to the version that is already pre-installed
with OS X.

Test everything that you have installed.  If your CPU is fast enough
(iMacs seem to do just fine), it will be able to play CDG files
smoothly using the native Python version, without requiring you to
build the optimised C version.  If, on the other hand, you observe a
lot of stuttering or sluggish graphic updates while playing CDG files,
then you will need to go a little bit further to compile the C
version.  This will require installing the XCode development suite,
and the SDL header files that match the version of SDL installed
automatically by pygame, above.  The details of this process are out
of the scope of this installation guide.

If you wish to play .avi files, .mov files, or .mpg files via an
external player, you will need to install a player that can be
launched via the command line, and which exits automatically when the
video is finished.  One such application is mplayer, which you can
install via MacPorts or fink.  There is also a precompiled version
available.  This is again beyond the scope of this installation guide.

---------------------------------------------------------------------------

INSTALLATION (WINDOWS)

Windows users can install PyKaraoke by simply downloading and running the
installer executable. This installs all prerequisite libraries, and adds
icons in your start menu to run PyKaraoke.

If you prefer, you may choose to build the Windows version from source. We
will assume you are familiar with the steps involved for installing a
Python distribution from source on Windows; they are similar to those for
the Linux installation, below. You will need to have pygame
(www.pygame.org) and wxpython (www.wxpython.org) installed. You will also
need to download and unpack the SDL source distribution (www.libsdl.org)
into the same directory with PyKaraoke, under its default name, which will
be something like "SDL-1.2.11" (by default, setup.py will search for any
directories named SDL* in the current directory). You will then invoke the
command:

python setup.py install

You may also choose to unpack the source distribution elsewhere and
specify its path with --sdl-location=/path/to/SDL on the above command.

---------------------------------------------------------------------------

CREATING WINDOWS INSTALLERS

To build a Windows installer, you additionally need to have py2exe
(www.py2exe.org) and NSIS (nsis.sourceforge.net) installed, and then you
may invoke:

python setup.py nsis

This will create an executable called pykaraoke-<current_version>.exe
in the current directory, which will be a standalone Windows installer
program you may then distribute to other Windows users. You don't
need to install PyKaraoke before building an installer for it.

---------------------------------------------------------------------------

INSTALLATION (MIDI/KAR FILE SUPPORT ON LINUX)

Windows users can enjoy MIDI/KAR file support using the standard
installation procedure.

MIDI/KAR support on Linux, however, requires the following:

 * Timidity++ (timidity.sourceforge.net)
 * Sound/patches for Timidity++

There are various sound patch collections available for Timidity++. Users
of PyKaraoke have used freepats and Eric Welsh's GUS patches.

To install Timidity++ on Gentoo together with Eric Welsh's patches use:

	# emerge timidity++ timidity-eawpatches

---------------------------------------------------------------------------

INSTALLATION (GP2X)

PyKaraoke can run on the GP2X handheld console, giving you a Karaoke
machine that fits in your pocket. You may find it easiest simply to
install the prebuilt binary version. The prebuilt version already
includes Python and the necessary supporting libraries. Simply unpack
this archive onto your SD card; it will expand to a small hierarchy of
files.  One of these directory names is pykaraoke/songs, which will be
empty; you should put your song files in this pykaraoke/songs
directory (or in any nested directory structure within
pykaraoke/songs). The first time you install, and each time you add
new song files to this directory, you should run the "rescan_songs"
script to rebuild the database with the new song files; but most of
the time, you should start PyKaraoke simply by running the script
named "pykaraoke".

As of PyKaraoke version 0.6, you can also play video karaoke files on
your GP2X.  You should use your existing GP2X tools to convert these
to AVI format, as supported by the GP2X version of mplayer, and put
them under the songs directory with your other karaoke files.
PyKaraoke will index them and present them along with all of your
other songs.

COMPILATION (GP2X)

If, for some reason, you wish to build your own version for the GP2X,
this is possible, but there will be a bit of work involved.  You will
need to install the GP2X development kit, which includes a
cross-compiler for the GP2X. You should next use this cross-compiler
to build and install SDL and PyGame, so that you have the appropriate
source headers and matching binaries for these library. Note it may be
necessary for you to apply patches to SDL_mixer to support using
tremor and libmad, which are alternative libraries used for playing
OGG and MP3 files, respectively.  (The default OGG and MP3
implementations used by SDL_mixer make heavy use of floating-point
arithmetic, which does not perform well on the GP2X.)  You will also
need to patch the timidity/config.h file to reduce the MIDI rendering
demands on the CPU. All of these patches are available for download
elsewhere.

To play AVI files, you will need a version of mplayer that accepts the
movie file on the command line (the default version supplied with the
GP2X firmware ignores the command-line arguments and always presents a
file-navigation GUI).  Use svn to obtain GPH's version of the mplayer
source code from:

http://svn.gp2x.com/gp2x/tag/application/2006.07.05/mplayer/

And then apply the patches given within this source archive in
install/mplayer-gp2x-cmdline-pykaraoke.diff .  This file combines the
patches available from the wiki at http://wiki.gp2x.org/wiki/MPlayer ,
but also adds a few fixes of my own (to fix relative avi files, and
also to automatically exit the player when the song has finished).

Once all that is done, you need to use the cross compiler to build
_pycdgAux.so. A sample script called cross-build-gp2x.sh is provided to do
this. Then it is simply a matter of copying this file, along with all of
the .py files, to the GP2X.

---------------------------------------------------------------------------

INSTRUCTIONS

If you used the install script you can start the player using:

	$ pykaraoke

Otherwise, start the player using:
	$ python pykaraoke.py


ChangeLog
*******************************************************************************
**** This is a part of PyKaraoke.                                          ****
****                                                                       ****
**** Copyright (C) 2007  Kelvin Lawson (kelvinl@users.sourceforge.net)     ****
**** Copyright (C) 2009  PyKaraoke Development Team                        ****
****                                                                       ****
**** This library is free software; you can redistribute it and/or         ****
**** modify it under the terms of the GNU Lesser General Public            ****
**** License as published by the Free Software Foundation; either          ****
**** version 2.1 of the License, or (at your option) any later version.    ****
****                                                                       ****
**** This library is distributed in the hope that it will be useful,       ****
**** but WITHOUT ANY WARRANTY; without even the implied warranty of        ****
**** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU     ****
**** Lesser General Public License for more details.                       ****
****                                                                       ****
**** You should have received a copy of the GNU Lesser General Public      ****
**** License along with this library; if not, write to the                 ****
**** Free Software Foundation, Inc.                                        ****
**** 59 Temple Place, Suite 330                                            ****
**** Boston, MA  02111-1307  USA                                           ****
*******************************************************************************

2011-08-23 Kelvin Lawson <kelvinl@users.sf.net>
 * Enhancements:
    pykdb.py: Reduce memory usage during duplicate file hash check.
    pykdb.py: Handle corrupted ZIP files.

2011-08-22 Kelvin Lawson <kelvinl@users.sf.net>
 * Bug fixes:
    pykdb.py: Fixes for unicode strings converted to ASCII.
	pykaraoke.py: Fixes for bytestrings that are not unicode.

2011-06-19 Kelvin Lawson <kelvinl@users.sf.net> 0.7.4
 * Bug fixes:
    pykaraoke.py: Allow search for unicode song names.
    pykaraoke.py: Fix FileDetails dialog for non-ascii characters.
    pykdb.py: Fixes for international characters in filenames.

2011-01-02 Kelvin Lawson <kelvinl@users.sf.net>
 * Enhancements:
    performer_prompt.py: Improve sizing on Windows.

2010-09-16 Kelvin Lawson <kelvinl@users.sf.net>
  * Enhancements:
    pykar.py: Better platform detection using ENV_POSIX.
    pykaraoke.py: Better platform detection using ENV_POSIX.
    pykconstants.py: Better platform detection using ENV_POSIX.
    pykenv.py: Better platform detection using ENV_POSIX.

2010-05-11 Kelvin Lawson <kelvinl@users.sf.net> 0.7.3
  * Bug fixes:
    -r1.62 pykaraoke.py: Make About window work with Wx2.6 again. Recent 
     About window changes work on Wx2.8 only.

2010-04-21
  * Enhancements:
     -r1.61 pykaraoke.py: Add options for double-buffered and hardware
     -r1.39 pykdb.py: Add options for double-buffered and hardware
     -r1.18 pykmanager.py: Add options for double-buffered and hardware
     -r1.12 pykplayer.py: 'P' key pauses.
     -r1.11 pykplayer.py: Backspace rewinds.

  * Bug fixes:
     - r1.9 _pcdgAux.c: Fix occasional crash in CDG player.

2010-04-20
  * Enhancements:
     -r1.60 pykaraoke.py: Support left/right margins, and two-column printing.
     -r1.20 pykar.py: OSX works better with Timidity.

2010-04-16
  * Enhancements:
     -r1.59 pykaraoke.py: Allow omitting filename from printout.

2010-04-14
  * Enhancements:
     -r1.38 pykdb.py: Remove articles from sort key, faster scrolling on GP2X.
     -r1.34 pycdg.py: Clean up keyboard operations and allow remote control
      access to pykaraoke_mini.
     -r1.10 pykaraoke_mini.py: Clean up keyboard operations and allow remote
      control access to pykaraoke_mini.
     -r1.9 pykaraoke_mini.py: Remove articles from sort key, faster scrolling on GP2X.
     -r1.19 pykar.py: Clean up keyboard operations and allow remote control
     -r1.17 pympg.py: Clean up keyboard operations and allow remote control

2010-04-09
  * Enhancements:
     -r1.8 pycdgAux.py: Filter out draw commands based on mysterious bit 0x20.
     -r1.8 _pycdgAux.c: Filter out draw commands based on mysterious bit 0x20.

2010-04-08
  * Enhancements:
     -r1.37 pykdb.py: Handle invalid utf-8 in titles.txt.

2010-03-31 Kelvin Lawson <kelvinl@users.sf.net> 0.7.3
  * Bug fixes:
    - r1.36 pykdb.py: Fix unicode folder searches. Add time string to temp
      filename to work around bug in pygame.mixer.music on Windows.
    - r1.33 pycdg.py: Fix unicode filename opens.
    - r1.18 pykar.py: Fix unicode filename opens.
    - r1.16 pympg.py: Fix unicode filename opens.

2010-03-29
  * Enhancements:
     -r1.58 pykaraoke.py: Minor OSX fixes

2009-10-25 Kelvin Lawson <kelvinl@users.sf.net> 0.7.2
  * Bug fixes:
    - r1.15 pympg.py: Support GetLength() call with external player. 

2009-08-26 Kelvin Lawson <kelvinl@users.sf.net> 0.7.1
  * Enhancements:
    - r1.57 pykaraoke.py: Improved layout of CDG Configuration page with
      song-derivation options.
    - r1.56 pykaraoke.py, r1.35 pykdb.py: Add new option to exclude filenames
      that don't match naming scheme from search results.
    - r1.33 pykdb.py: Move from md5 to hashlib due to Python deprecation
      warning.
    - r1.32 pykdb.py: Now allows songs in database if they do not match the 
      "Derive song information from filename" format. Sets the title field
      only to the filename.
    - r1.31 pycdg.py: Switch from MP3Info.py to python-mutagen for MP3 info
      as more widely available in package managers.

  * Bug Fixes:
    - r1.31 pykdb.py: Fix failure to play songs from File View due to addition 
      of "Derive song information from filename" feature.

pykplayer.py
#
# Copyright (C) 2010  Kelvin Lawson (kelvinl@users.sourceforge.net)
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

"""This module defines the class pykPlayer, which is a base class used
by the modules pykar.py, pycdg.py, and pympg.py.  This collects
together some common interfaces used by these different
implementations for different types of Karaoke files."""

from pykconstants import *
from pykmanager import manager
from pykenv import env
import pygame, wx
import sys
import pygst, gst
import types
import os

class pykPlayer:
    def __init__(self, song, songDb,
                 errorNotifyCallback = None, doneCallback = None,
                 windowTitle = None):
        """The first parameter, song, may be either a pykdb.SongStruct
        instance, or it may be a filename. """

        if songDb == None:
            import pykdb
            songDb = pykdb.globalSongDB
            songDb.LoadSettings(None)
        self.songDb = songDb

        # Set the global command-line options if they have not already
        # been set.
        if manager.options == None:
            parser = self.SetupOptions()
            (manager.options, args) = parser.parse_args()
            manager.ApplyOptions(self.songDb)

            if song is None:
                if (len(args) != 1):
                    parser.print_help()
                    sys.exit(2)
                song = args[0]

        # Unfortunately, we can't capture sound when dumping.  There
        # are two reasons for this.  (1) pymedia doesn't currently
        # support multiplexing audio with a video stream, so when
        # you're dumping an mpeg file, it has to be video-only.  (2)
        # pygame doesn't provide a way for us to programmatically
        # convert a midi file to sound samples anyway--all you can do
        # with a midi file is route it through the speakers.

        # So, for these reasons, we always just disable sound when
        # dumping images or movies.
        if manager.options.dump:
            manager.options.nomusic = True

        if isinstance(song, types.StringTypes):
            # We were given a filename.  Convert it to a SongStruct.
            song = self.songDb.makeSongStruct(song)
        
        # Store the parameters
        self.Song = song
        self.WindowTitle = windowTitle

        # And look up the actual files corresponding to this SongStruct.
        self.SongDatas = song.GetSongDatas()
        if windowTitle is None:
            self.WindowTitle = song.DisplayFilename
            
        # Caller can register a callback by which we
        # print out error information, use stdout if none registered
        if errorNotifyCallback:
            self.ErrorNotifyCallback = errorNotifyCallback
        else:
            self.ErrorNotifyCallback = self.__defaultErrorPrint
    
        # Caller can register a callback by which we
        # let them know when the song is finished
        if doneCallback:
            self.SongFinishedCallback = doneCallback
        else:
            self.SongFinishedCallback = None

        self.State = STATE_INIT
        self.InternalOffsetTime = 0

        # These values are used to keep track of the current position
        # through the song based on pygame's get_ticks() interface.
        # It's used only when get_pos() cannot be used or is
        # unreliable for some reason.
        self.PlayTime = 0
        self.PlayStartTime = 0
        self.PlayFrame = 0

        # self.PlayStartTime is valid while State == STATE_PLAYING; it
        # indicates the get_ticks() value at which the song started
        # (adjusted for any pause intervals that occurred during
        # play).  self.PlayTime is valid while State != STATE_PLAYING;
        # it indicates the total number of ticks (milliseconds) that
        # have elapsed in the song so far.

        # self.PlayFrame starts at 0 and increments once for each
        # frame.  It's not very meaningful, except in STATE_CAPTURING
        # mode.

        # Keep track of the set of modifier buttons that are held
        # down.  This is currently used only for the GP2X interface.
        self.ShoulderLHeld = False
        self.ShoulderRHeld = False

        # Set this true if the player can zoom font sizes.
        self.SupportsFontZoom = False

    # The following methods are part of the public API and intended to
    # be exported from this class.

    def Validate(self):
        """ Returns True if the karaoke file appears to be playable
        and contains lyrics, or False otherwise. """

        return self.doValidate()

    def Play(self):
        self.doPlay()

        if manager.options.dump:
            self.setupDump()
        else:
            self.PlayStartTime = pygame.time.get_ticks()
            self.State = STATE_PLAYING

    # Pause the song - Use Pause() again to unpause
    def Pause(self):
        if self.State == STATE_PLAYING:
            self.doPause()
            self.PlayTime = pygame.time.get_ticks() - self.PlayStartTime
            self.State = STATE_PAUSED
        elif self.State == STATE_PAUSED:
            self.doUnpause()
            self.PlayStartTime = pygame.time.get_ticks() - self.PlayTime
            self.State = STATE_PLAYING

    # Close the whole thing down
    def Close(self):
        self.ClearHostLyrics()
        self.State = STATE_CLOSING
        manager.Music.set_state(gst.STATE_NULL)

    # you must call Play() to restart. Blocks until pygame is initialised
    def Rewind(self):
        self.doRewind()
        self.PlayTime = 0
        self.PlayStartTime = 0
        self.PlayFrame = 0
        self.State = STATE_NOT_PLAYING

    # Stop the song and go back to the start. As you would
    # expect Stop to do on a CD player. Play() restarts from
    # the beginning
    def Stop(self):
        self.ClearHostLyrics()
        manager.Music.set_state(gst.STATE_NULL)
        self.Rewind()
            
    # Get the song length (in seconds)
    def GetLength(self):
        ErrorString = "GetLength() not supported"
        self.ErrorNotifyCallback (ErrorString)
        return None

    # Get the current time (in milliseconds).
    def GetPos(self):
        if self.State == STATE_PLAYING:
            return pygame.time.get_ticks() - self.PlayStartTime
        else:
            return self.PlayTime

    def SetupOptions(self, usage = None):
        """ Initialise and return optparse OptionParser object,
        suitable for parsing the command line options to this
        application. """

        if usage == None:
            usage = "%prog [options] <Karaoke file>"

        return manager.SetupOptions(usage, self.songDb)

    # Below methods are internal.

    def setupDump(self):
        # Capture the output as a sequence of numbered frame images.
        self.PlayTime = 0
        self.PlayStartTime = 0
        self.PlayFrame = 0
        self.State = STATE_CAPTURING

        self.dumpFrameRate = manager.options.dump_fps
        assert self.dumpFrameRate

        filename = manager.options.dump
        base, ext = os.path.splitext(filename)
        ext_lower = ext.lower()

        self.dumpEncoder = None
        if ext_lower == '.mpg':
            # Use pymedia to convert frames to an mpeg2 stream
            # on-the-fly.
            import pymedia
            import pymedia.video.vcodec as vcodec

            self.dumpFile = open(filename, 'wb')
            frameRate = int(self.dumpFrameRate * 100 + 0.5)
            self.dumpFrameRate = float(frameRate) / 100.0
            
            params= { \
              'type': 0,
              'gop_size': 12,
              'frame_rate_base': 125,
              'max_b_frames': 0,
              'height': manager.options.size_y,
              'width': manager.options.size_x,
              'frame_rate': frameRate,
              'deinterlace': 0,
              'bitrate': 9800000,
              'id': vcodec.getCodecID('mpeg2video')
            }
            self.dumpEncoder = vcodec.Encoder( params )
            return
            
        # Don't dump a video file; dump a sequence of frames instead.
        self.dumpPPM = (ext_lower == '.ppm' or ext_lower == '.pnm')
        self.dumpAppend = False
        
        # Convert the filename to a pattern.
        if '#' in filename:
            hash = filename.index('#')
            end = hash
            while end < len(filename) and filename[end] == '#':
                end += 1
            count = end - hash
            filename = filename[:hash] + '%0' + str(count) + 'd' + filename[end:]
        else:
            # There's no hash in the filename.
            if self.dumpPPM:
                # We can dump a series of frames all to the same file,
                # if we're dumping ppm frames.  Mjpegtools likes this.
                self.dumpAppend = True
                try:
                    os.remove(filename)
                except OSError:
                    pass
            else:
                # Implicitly append a frame number.
                filename = base + '%04d' + ext
            
        self.dumpFilename = filename

    def doFrameDump(self):
        if self.dumpEncoder:
            import pymedia.video.vcodec as vcodec

            ss = pygame.image.tostring(manager.surface, "RGB")
            bmpFrame = vcodec.VFrame(
                vcodec.formats.PIX_FMT_RGB24,
                manager.surface.get_size(), (ss,None,None))
            yuvFrame = bmpFrame.convert(vcodec.formats.PIX_FMT_YUV420P)
            d = self.dumpEncoder.encode(yuvFrame)
            self.dumpFile.write(d.data)
            return

        if self.dumpAppend:
            filename = self.dumpFilename
        else:
            filename = self.dumpFilename % self.PlayFrame
            print filename

        if self.dumpPPM:
            # Dump a PPM file.  We do PPM by hand since pygame
            # doesn't support it directly, but it's so easy and
            # useful.
            
            w, h = manager.surface.get_size()
            if self.dumpAppend:
                f = open(filename, 'ab')
            else:
                f = open(filename, 'wb')
            f.write('P6\n%s %s 255\n' % (w, h))
            f.write(pygame.image.tostring(manager.surface, 'RGB'))

        else:
            # Ask pygame to dump the file.  We trust that pygame knows
            # how to store an image in the requested format.
            pygame.image.save(manager.surface, filename)
            

    def doValidate(self):
        return True
    
    def doPlay(self):
        pass

    def doPause(self): 
        pass

    def doUnpause(self):
        pass

    def doRewind(self):
        pass

    def doStuff(self):
        # Override this in a derived class to do some useful per-frame
        # activity.
        # Common handling code for a close request or if the
        # pygame window was quit
        if self.State == STATE_CLOSING:
            if manager.display:
                manager.display.fill((0,0,0))
                pygame.display.flip()
            self.shutdown()

        elif self.State == STATE_CAPTURING:
            # We are capturing a video file.
            self.doFrameDump()

            # Set the frame time for the next frame.
            self.PlayTime = 1000.0 * self.PlayFrame / self.dumpFrameRate
            
        self.PlayFrame += 1
        

    def doResize(self, newSize):
        # This will be called internally whenever the window is
        # resized for any reason, either due to an application resize
        # request being processed, or due to the user dragging the
        # window handles.
        pass

    def doResizeBegin(self):
        # This will be called internally before the screen is resized
        # by pykmanager and doResize() is called. Not all players need
        # to do anything here.
        pass

    def doResizeEnd(self):
        # This will be called internally after the screen is resized
        # by pykmanager and doResize() is called. Not all players need
        # to do anything here.
        pass

    def ClearHostLyrics(self):
        # Update the host lyrics
        if manager.HostLyrics:
            manager.HostLyrics.SetBitmap(manager.HostSplash)

    def UpdateHostLyrics(self):
        # Update the host lyrics
        if manager.HostLyrics:
            hostscaled = pygame.transform.smoothscale(manager.surface, (160, 120))
            IMG = wx.EmptyImage(160, 120)
            IMG.SetData(pygame.image.tostring(hostscaled, 'RGB'))
            manager.HostLyrics.SetBitmap(IMG.ConvertToBitmap())
            hostscaled.unlock()

    def handleEvent(self, event):
        if event.type == pygame.USEREVENT:
            self.Close()
        elif event.type == pygame.KEYDOWN:
            if event.key == pygame.K_ESCAPE:
                self.Close()

            elif event.key == pygame.K_PAUSE or event.key == pygame.K_p:
                self.Pause()

            elif event.key == pygame.K_BACKSPACE or event.key == pygame.K_DELETE:
                self.Rewind()
                self.Play()

            # Use control-left/right arrow to offset the current
            # graphics time by 1/4 sec.  Use control-down arrow to
            # restore them to sync.
            elif self.State == STATE_PLAYING and event.key == pygame.K_RIGHT and event.mod & (pygame.KMOD_LCTRL | pygame.KMOD_RCTRL):
                manager.settings.SyncDelayMs += 250
                print "sync %s" % manager.settings.SyncDelayMs
            elif self.State == STATE_PLAYING and event.key == pygame.K_LEFT and event.mod & (pygame.KMOD_LCTRL | pygame.KMOD_RCTRL):
                manager.settings.SyncDelayMs -= 250
                print "sync %s" % manager.settings.SyncDelayMs
            elif self.State == STATE_PLAYING and event.key == pygame.K_DOWN and event.mod & (pygame.KMOD_LCTRL | pygame.KMOD_RCTRL):
                manager.settings.SyncDelayMs = 0
                print "sync %s" % manager.settings.SyncDelayMs

            if self.SupportsFontZoom:
                if event.key == pygame.K_PLUS or event.key == pygame.K_EQUALS or \
                   event.key == pygame.K_KP_PLUS:
                    manager.ZoomFont(1.0/0.9)
                elif event.key == pygame.K_MINUS or event.key == pygame.K_UNDERSCORE or \
                   event.key == pygame.K_KP_MINUS:
                    manager.ZoomFont(0.9)

        elif event.type == pygame.QUIT:
            self.Close()
            
        elif env == ENV_GP2X and event.type == pygame.JOYBUTTONDOWN:
            if event.button == GP2X_BUTTON_SELECT:
                self.Close()
            elif event.button == GP2X_BUTTON_START:
                self.Pause()
            elif event.button == GP2X_BUTTON_L:
                self.ShoulderLHeld = True
            elif event.button == GP2X_BUTTON_R:
                self.ShoulderRHeld = True

            if self.SupportsFontZoom:
                if event.button == GP2X_BUTTON_RIGHT and self.ShoulderLHeld:
                    manager.ZoomFont(1.0/0.9)
                elif event.button == GP2X_BUTTON_LEFT and self.ShoulderLHeld:
                    manager.ZoomFont(0.9)
            
        elif env == ENV_GP2X and event.type == pygame.JOYBUTTONUP:
            if event.button == GP2X_BUTTON_L:
                self.ShoulderLHeld = False
            elif event.button == GP2X_BUTTON_R:
                self.ShoulderRHeld = False

    def shutdown(self):
        # This will be called by the pykManager to shut down the thing
        # immediately.
        self.ClearHostLyrics()
        manager.Music.set_state(gst.STATE_NULL)

        # If the caller gave us a callback, let them know we're finished
        if self.State != STATE_CLOSED:
            self.State = STATE_CLOSED
            if self.SongFinishedCallback != None:
                self.SongFinishedCallback()


    def __defaultErrorPrint(self, ErrorString):
        print (ErrorString)


    def findPygameFont(self, fontData, fontSize):
        """ Returns a pygame.Font selected by this data. """
        if not fontData.size:
            # The font names a specific filename.
            filename = fontData.name
            if os.path.sep not in filename:
                filename = os.path.join(manager.FontPath, filename)
            return pygame.font.Font(filename, fontSize)

        # The font names a system font.
        pointSize = int(fontData.size * fontSize / 10.0 + 0.5)
        return pygame.font.SysFont(
            fontData.name, pointSize, bold = fontData.bold,
            italic = fontData.italic)
